apm-agent-core/src/main/java/co/elastic/apm/agent/util/BinaryHeaderMap.java

```diff
@@ -0,0 +1,223 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.util;
+
+import co.elastic.apm.agent.objectpool.Recyclable;
+import co.elastic.apm.agent.report.serialize.DslJsonSerializer;
+
+import javax.annotation.Nullable;
+import java.nio.Buffer;
+import java.nio.CharBuffer;
+import java.nio.charset.CoderResult;
+import java.util.ArrayList;
+import java.util.Iterator;
+
+/**
+ * A map (not implementing the java.util.Map interface) that only supports String-byte[] pair additions and iterations.
+ * The map doesn't allocate during addition or iteration.
+ * This map does not support any form of concurrency. It can be either be in a write mode (through its {@link #add}
+ * method) or read mode (through the {@link #iterator()} API) at a given time. Reads and writes must not be
+ * performed by more than one thread concurrently.
+ * <p>
+ * NOTE: this map does not guarantee visibility, therefore ensuring visibility when switching from read to write mode
+ * (or the other way around) is under the responsibility of the map's user.
+ */
+public class BinaryHeaderMap implements Recyclable, Iterable<BinaryHeaderMap.Entry> {
+    public static final int MAXIMUM_HEADER_BUFFER_SIZE = DslJsonSerializer.MAX_VALUE_LENGTH * 10;
+
+    private CharBuffer valueBuffer;
+    private final ArrayList<String> keys;
+    private int[] valueLengths;
+    private final NoGarbageIterator iterator;
+
+    public BinaryHeaderMap() {
+        valueBuffer = CharBuffer.allocate(64);
+        keys = new ArrayList<>(10);
+        valueLengths = new int[10];
+        iterator = new NoGarbageIterator();
+    }
+
+    public int size() {
+        return keys.size();
+    }
+
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    public boolean add(String key, byte[] value) throws InsufficientCapacityException {
+        int valuesPos = valueBuffer.position();
+        CoderResult coderResult = IOUtils.decodeUtf8Bytes(value, valueBuffer);
+        while (coderResult.isOverflow()) {
+            ((Buffer) valueBuffer).limit(valuesPos);
+            enlargeBuffer();
+            coderResult = IOUtils.decodeUtf8Bytes(value, valueBuffer);
+        }
+        boolean result;
+        if (coderResult.isError()) {
+            ((Buffer) valueBuffer).limit(valuesPos);
+            result = false;
+        } else {
+            int size = keys.size();
+            keys.add(key);
+            if (size == valueLengths.length) {
+                enlargeValueLengths();
+            }
+            valueLengths[size] = valueBuffer.position() - valuesPos;
+            result = true;
+        }
+        return result;
+    }
+
+    private void enlargeBuffer() throws InsufficientCapacityException {
+        if (valueBuffer.capacity() == MAXIMUM_HEADER_BUFFER_SIZE) {
+            throw new InsufficientCapacityException();
```

Comment:
Avoid creating a stack trace with the singleton exception pattern. See also `co.elastic.apm.agent.bci.bytebuddy.MinimumClassFileVersionValidator.UnsupportedClassFileVersionException`.

Reply:
Another approach would be to return a boolean